aggregator/list.go

```diff
@@ -688,11 +688,87 @@ func (l *forwardedMetricList) Close() {
 	}
 }
 
+type timedMetricListID struct {
+	resolution time.Duration
+}
+
+func (id timedMetricListID) toMetricListID() metricListID {
+	return metricListID{listType: timedMetricListType, timed: id}
+}
+
+type timedMetricList struct {
+	*baseMetricList
+
+	flushOffset time.Duration
+	log         log.Logger
+	flushMgr    FlushManager
+}
+
+func newTimedMetricList(
+	shard uint32,
+	id timedMetricListID,
+	opts Options,
+) (*timedMetricList, error) {
+	var (
+		resolution                 = id.resolution
+		fn                         = opts.BufferForPastTimedMetricFn()
+		timedAggregationBufferPast = fn(resolution)
+		iOpts                      = opts.InstrumentOptions()
+		listScope                  = iOpts.MetricsScope().Tagged(map[string]string{"list-type": "timed"})
+	)
+	// Timed metrics that have been kept for longer than the maximum buffer
+	// will be flushed.
+	targetNanosFn := func(nowNanos int64) int64 {
+		return nowNanos - timedAggregationBufferPast.Nanoseconds()
+	}
+	l, err := newBaseMetricList(
+		shard,
+		resolution,
+		targetNanosFn,
+		isStandardMetricEarlierThan,
+		standardMetricTimestampNanos,
+		opts.SetInstrumentOptions(iOpts.SetMetricsScope(listScope)),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	flushOffset := timedAggregationBufferPast - timedAggregationBufferPast.Truncate(l.FlushInterval())
+	fl := &timedMetricList{
+		baseMetricList: l,
+		flushOffset:    flushOffset,
+		log:            opts.InstrumentOptions().Logger(),
+		flushMgr:       opts.FlushManager(),
+	}
+	fl.flushMgr.Register(fl)
+	return fl, nil
+}
+
+func (l *timedMetricList) ID() metricListID {
+	return timedMetricListID{
+		resolution: l.resolution,
+	}.toMetricListID()
+}
+
+func (l *timedMetricList) FlushOffset() time.Duration {
```

Comment:
Do the timed metrics need to be flushed at a fixed offset like the forwarded ones?